R/slice.R

```diff
@@ -165,6 +174,81 @@ vec_as_index <- function(i, n, names = NULL) {
   vec_assert(n, integer(), 1L)
   .Call(vctrs_as_index, i, n, names)
 }
+#' @rdname vec_as_index
+#' @export
+vec_as_position <- function(i, n, names = NULL) {
+  if (is.object(i)) {
+    if (vec_is_coercible(i, int())) {
+      i <- vec_cast(i, int())
+    } else if (vec_is_coercible(i, chr())) {
+      i <- vec_cast(i, chr())
+    }
+  } else if (typeof(i) == "double") {
+    i <- vec_coercible_cast(i, int())
+  }
+
+  type <- typeof(i)
+  if (!type %in% c("integer", "character")) {
+    stop_position_incompatible_type(i)
+  }
+  if (length(i) != 1L) {
+    stop_position_incompatible_size(i)
+  }
+  if (type == "integer" && i < 1L) {
+    stop_position_incompatible_sign(i)
+  }
+
+  vec_as_index(i, n, names = names)
+}
+
+stop_position_incompatible_type <- function(i) {
```

Comment:
Do we really want one error class for each of this? It feels very fine grained

Reply:
Thinking about the error messages:

```R
Error: Invalid type in argument `i`
* `i` is size 2
* `i` must be size 1

Error: Invalid type in argument `i`
* `i` is a logical vector
* `i` must be a character or integer vector

Error: Invalid value in argument `i`
* `i` is -1
* `i` must be greater than 0
```

I'm not sure the distinction between value and type is sufficient to require a different class.

Reply:
I agree it feels too fine-grained, but to merge these error classes we need:

* A concept that includes both type and size. I like the idea of working around this with the notion of invalid value.

* Perform the type checking again as part of the condition-message method. While a bit redundant, this also means the `stop_` constructor is useful in more cases. For instance in the case of vectorised errors, it can be called from a fast C routine that doesn't keep track of the bad values.

I will revisit.

What do you think about using `bad` instead of `invalid`? It makes for shorter error classes and messages.

Reply:
oh or were you suggesting making size errors a subset of type errors?

Reply:
I was thinking something like:

```R
stop_bad_argument("i", vector_of_problems)
```

But I don't see how to move the logic into the `conditionMessage()` method

Reply:
I wondered about that as well, maybe we could come up with a new generic that generates error bullets. Then the problems need to be wrapped up in objects, and probably error conditions are most suitable. The condition-message method for these conditions would invoke the bullet generic as well.